src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/StackWalk/Context/Unwinder.cs

```
@@ -109,7 +109,15 @@ private static unsafe int ReadFromTarget(ulong address, void* pBuffer, int buffe
             return -1;
```

(Existing code, so not strictly your problem)
This `GCHandle.FromIntPtr` on `context` makes me suspicious, doubly so that we're checking whether the target is some type other than what we expect. Are we trying to guard against garbage context pointers here? Or something else? A comment might be good saying what this if statement is solving for people who look at the code later, if you happen to know. If it's equally mysterious to you then you can put it out of your mind and resolve this comment :)

---

Looking through the existing code it looks like this is an abstract unwinder native API that lets you smuggle arbitrary data through `void* context`, so since we're providing an IntPtr when calling it elsewhere in this file, it's probably fine? But in that case, I'd like to see us fail fast and fail loudly - we shouldn't return -1, we should probably bail out and throw an exception because somehow we passed ourselves something other than the context we expected. Maybe I'm not understanding this.

---

No totally reasonable. I also wrote that code. 

This is guarding against memory corruption / an invalid call. The intptr should always be a GCHandle with Target ContextCallback. I can add a comment.

---

OK. If we're concerned about memory corruption, I'm not certain that it's safe to turn arbitrary bytes into a GCHandle and dereference it. Hopefully we bounds-check the table lookup? I'll see if I can find out...

---

`GCHandle.ThrowIfInvalid` throws for `0`, so a `null` `context` would make us throw here instead of returning `-1`.
`GCHandleInternalGet` does not appear to do any verification or bounds checking. It calls `ObjectFromHandle`, which also does not appear to do any verification or bounds checking. So I think if we got something here that wasn't a GCHandle, Anything Could Happen. Someone who knows the VM better could correct me here.

---

One of the issues is that this is a callback from a native code boundary. Therefore, we probably shouldn't throw because on non-Windows platforms that's a fail fast. Ideally, we'd return an error to the native code which would propagate that error back to the managed invoking function.